src/supervisor.tex

```diff
@@ -1440,11 +1516,14 @@ \subsection{Virtual Address Translation Process}
 \begin{enumerate}
 
 \item Let $a$ be ${\tt satp}.ppn \times \textrm{PAGESIZE}$, and let $i=\textrm{LEVELS} - 1$. (For Sv32, PAGESIZE=$2^{12}$ and LEVELS=2.)
+  The {\tt satp} register must be {\em active}, i.e., the effective privilege
+  mode must be S-mode or U-mode.
 
 \item Let $pte$ be the value of the PTE at address
   $a+va.vpn[i]\times \textrm{PTESIZE}$. (For Sv32, PTESIZE=4.)
-  If accessing $pte$ violates a PMA or PMP check, raise an
-  access-fault exception corresponding to the original access type.
+  If accessing $pte$ violates a PMA or PMP check, or if any reserved
+  bits in $pte$ are set, raise an access-fault exception corresponding to the
```

Comment:
Why does setting reserved bits in a PTE cause an access fault instead of a page fault? What will M-mode software do with this?

Seems a bit inconsistent to me since all other PTE issues end up as page faults, which can easily be delegated to S-mode.

Reply:
To paraphrase a prior discussion with @aswaterman, invalid use of defined VM features results in a page fault, while invalid physical addresses cause access faults.  Making invalid use of reserved bits trigger access faults also ensures that the behavior would be consistent if any reserved bits are later repurposed as added PPN bits.

Invalid use of NAPOT PTEs (now on #628) is a page fault because it's invalid use of a defined feature, and it is also consistent with the way misaligned superpages are handled.

In some cases different decisions could have been made, but that's the reasoning for the current choices.

Reply:
This is the first instance (I think) where the translation mechanism generates an access fault when the PTE is not well-formed (i.e. the reserved bits are non-zero).

The model seems simpler and more modular if the translation mechanism checks the PTEs for proper addresses and permissions and then hands off valid, well-formed PAs to be checked by the PMAs and PMPs. Conversely, if an access raises a page fault, software knows something was wrong with the translation process and data structures, but if an access raises an access fault, software knows that the fault lies somewhere besides the translation process and data structures.

Reply:
It's then inconsistent if you repurpose the bits later to be flags (e.g. as is done for NAPOT). You can't win either way, but making it be a page fault is more natural than pretending they're address bits but actually aren't, and keeps the semantics from the existing spec version, which I think should always be done unless there's a very good reason why it _can't_ be.

Reply:
I'll forward this question to the task group for more discussion.

IIRC in earlier discussions we felt that the existing spec wasn't actually unambiguously clear about this case.  There is wording saying "Bits 63–54 are reserved for future standard use and must be zeroed by software for forward compatibility", but there isn't actually any explicit wording saying what happens if that is violated.  It doesn't show up explicitly in the translation algorithm for example.  It seemed like we needed to actually make a decision one way or the other in order to clarify a hole in the spec.  It wasn't the intention to change existing semantics.

@jrtc27 can you point to the pre-existing spec wording I'm missing that makes this clearly a page fault?  If so, then I agree we should stick with that.  If not, then we should just decide which is better and formalize it.

FWIW, spike also doesn't seem to check for this either way, while the Sail model considers this a page fault.

Reply:
Hm, I could have sworn it was there, but you're right, I can't find it. All the implementations I know of either ignore the bits or raise a page fault, though, and I do think a page fault would be more natural.

Reply:
In our discussions it was pointed out that Rocket treats this as an access fault rather than a page fault: https://github.com/chipsalliance/rocket-chip/blob/8f9149faf01fec9725d08c31c9c0bb5341f5e999/src/main/scala/rocket/PTW.scala#L154

So unfortunately we seem to have both behaviors in the wild already.

Reply:
After discussing, we have indeed changed this to a page fault.  The expectation is that all misconfigurations of a page table entry (use of reserved bits, misaligned superpages, invalid use of r/w/x, invalid use of Svnapot) will result in page faults, including if/when we start using more reserved bits for other purposes.  Access faults will remain for things like unsupported misaligned accesses or invalid physical addresses, which aren't specific to the translation process.

Note that this rule will apply once priv 1.12 is ratified.  It is not a retroactive change, as implementations raising access faults exist in the wild.

Thanks all for the feedback!

Reply:
tracking Rocket PTW change request: https://github.com/chipsalliance/rocket-chip/issues/2795

Reply:
If an implementation today raises a fault when reserved PTE bits are used, then it'll be incompatible with future software making use of those bits. Hw should just ignore reserved bits, or else we 'll need to fill the sw with checks on the critical path so that we choose PTE formats based on priv. spec revisions, or chose PTE formats at compile-time (meaning we can't have one kernel image for all systems).

It also seems a bit inconsistent to me, for CSRs for example WPRI fields ignore writes to reserved bits (hardwired to 0), WARL fields also ignore them ("Implementations will not raise an exception") and WLRL fields "are permitted but not required to raise an exception". If we write an invalid/reserved SATP value, it's ignored, if we use an invalid/reserved bit on sfence.vma 's rs2 it's ignored, and TBH I couldn't find a case on the current Priv. spec where writing a reserved bit/field requires the hw to raise an exception. Another inconsistency is that there are not just reserved PTE bits but also reserved combinations of standard PTEs, e.g. RW=01, or reserved combinations of D,A,U bits on non-leaf PTEs. If we are about to return a fault for invalid PTEs, we should include those cases as well.

Reply:
This was already debated in the discussion above and in the virtual memory TG.  As of Priv 1.12, once this change is ratified, any invalid encoding (use of upper reserved bits, RW=01, reserved combinations of D,A,U) will trigger a page fault.  Plus, as discussed above, most implementations already raise a page fault or an access fault, and we'd like to make this consistent (as a page fault) moving forward.  Is there new information here that should cause us to open this up again?

What new checks would have to be performed on the critical path?  The OS obviously shouldn't attempt to use the new features if they aren't supported by a given implementation, and in both extensions we've proposed so far the zero(s) encoding is backwards compatible with the existing behavior.

Reply:
I get it that invalid use of _defined_ VM features should result a fault, but we are talking about features to be defined at a later spec revision, using reserved bits currently unused. Why should we raise an exception on those where we can just ignore them ? And what is sw supposed to do in this case ? We won't even be able to call the trap handler to print the crash info, we'll just hang. We are turning this case to a fatal error for no apparent reason (the hw can work fine by ignoring those bits). As for the critical path and the changes needed, on Linux the core page-table related functions are defined on header files (in our case arch/riscv/include/asm/pgtable.h) so that they can be inlined, and the idea is to keep them as simple as possible (we do mappings all the time, this needs to be as efficient as possible), if we start adding checks there for every new PTE bit we decide to add in the future e.g. on set_pte_at we'll end up having something like powerpc's set_pte_filter (https://elixir.bootlin.com/linux/latest/source/arch/powerpc/mm/pgtable.c#L103) only worse (and that's the only arch I could find that follows this approach, others just use a different header file at compile time, something we want to avoid since we want to have one image that runs everywhere). I'm not saying it's not possible or wrong, I'm saying it's a bit ugly and sub-optimal sw-wise.

As for the consistency, the text still says "For non-leaf PTEs, the D, A, and U bits are reserved for future standard use and must be cleared by software for forward compatibility.", also "Writable pages must also be marked readable; the contrary combinations are reserved for future use", shouldn't we also say that a page fault must occur in those cases like we do for reserved PTE bits on Sv39 ? It's only mentioned on the translation process algorithm.